### PR TITLE
SNOW-1926346: Implement groupby transform without pivot

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1926346

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [ ] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)

3. Please describe how your code solves the related issue.

   This PR reimplements groupby.transform which does not use pivot. Output of transform maintains same row index as existing dataframe and column index omits by columns. We use this property of transform to correctly create output schema of UDTF and avoid expensive pivot step.
   Note: We still generate VARIANT columns (same as existing implementation). In follow up PR use type inference framework to return correct types in output.
   Note: It's possible that lambda passed in transform does not adhere to expected output schema. It should be rare but will improve this as well when we add type inference in groupby.transform.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 1]
```diff
@@ -568,8 +794,6 @@ def end_partition(self, df: native_pd.DataFrame):  # type: ignore[no-untyped-def
             A dataframe representing the result of applying the user-provided
             function to this group.
             """
-            current_column_position = 0
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:571]: removed dead code.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -4126,12 +4129,10 @@ def groupby_apply(
             ],
             existing_identifiers=self._modin_frame.ordered_dataframe._dataframe_ref.snowflake_quoted_identifiers,
             force_list_like_to_series=force_list_like_to_series,
+            is_transform=is_transform,
         )
 
         new_internal_df = self._modin_frame.ensure_row_position_column()
-        row_position_snowflake_quoted_identifier = (
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:4132]: removed dead code.

**Comment by sfc-gh-jjiao** [comment]
Thanks @sfc-gh-nkumar for improving groupby. Is there any benchmarking and maybe manual testing result showing the performance improvement?

**Comment by sfc-gh-nkumar** [comment]
> Thanks @sfc-gh-nkumar for improving groupby. Is there any benchmarking and maybe manual testing result showing the performance improvement?

Local testing shows reduction from 66 sec to 26 sec on microbenchmark. Planning to add bunch of these to framework to make this tracking permanent.

**Comment by sfc-gh-jjiao** [comment]
1. Thanks for the PR with informative description.  QQ:  the description said " It's possible that lambda passed in transform does not adhere to expected output schema. It should be rare", what is the behavior when the lambda passed does not adhere to expected output schema?
2. why all the precommit tests failed?

**Review by sfc-gh-jkew — APPROVED** [review]
This looks pretty good to me. What happens with `test_dataframe_groupby_transform_conflicting_labels_negative` - does it just not throw an exception or what?

**Comment by sfc-gh-helmeleegy** [comment]
> 3\. Note: It's possible that lambda passed in transform does not adhere to expected output schema. It should be rare but will improve this as well when we add type inference in groupby.transform.

I think it would be good if we have a test to exercise this case.

**Comment by sfc-gh-helmeleegy** [comment]
> 1. Thanks for the PR with informative description.  QQ:  the description said " It's possible that lambda passed in transform does not adhere to expected output schema. It should be rare", what is the behavior when the lambda passed does not adhere to expected output schema?
> 2. why all the precommit tests failed?

For #2, the one failing Modin test should pass after merging with main. The rest are non-modin/pandas related, and they are not required to pass either.

File: src/snowflake/snowpark/modin/plugin/extensions/groupby_overrides.py [hunk 3]
```diff
@@ -130,6 +130,9 @@ def __init__(
             "group_keys": group_keys,
         }
         self._kwargs.update(kwargs)
+        if "apply_op" not in self._kwargs:
+            # Can be "apply", "transform", "filter" or "aggregate"
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/extensions/groupby_overrides.py:134]: Do we have tests for each of these functions?

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 4]
```diff
@@ -434,6 +434,255 @@ def create_groupby_transform_func(
...
+        # For SeriesGroupBy, there should be only one data column.
+        num_columns = len(df.columns)
+        assert (
+            num_columns == 1
+        ), f"Internal error: SeriesGroupBy func should apply to series, but input data had {num_columns} columns."
+        input_object = df.iloc[:, 0].rename(group_label)
+    else:
+        input_object = df.set_axis(data_column_index, axis="columns")
+    # Use infer_objects() because integer columns come as floats
+    # TODO: file snowpark bug about that. Asked about this here:
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:506]: I know the TODO here is for filing a bug. But the convention has been to include the jira ticket along with the TODO comment. So wondering if you can actually file the bug and include it here?

File: tests/integ/modin/groupby/test_groupby_transform.py [hunk 5]
```diff
@@ -141,9 +144,9 @@ def test_dataframe_groupby_transform_conflicting_labels_negative():
 
 
 @sql_count_checker(
-    query_count=11,
-    join_count=8,
-    udtf_count=2,
+    query_count=7,
+    join_count=12,
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_transform.py:148]: Can we have a comment explaining the reason for the high join count?

File: tests/integ/modin/groupby/test_groupby_transform.py [hunk 6]
```diff
@@ -112,6 +112,9 @@ def test_dataframe_groupby_transform_with_func_args_and_kwargs(
         )
 
 
+@pytest.mark.skip(
+    reason="SNOW-1933703: Raise NotImplementedError for groupby transform"
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/groupby/test_groupby_transform.py:116]: This is a bit confusing. The test is already checking for raising a NotImplementedError. Not clear why we needed to skip it?

**Comment by sfc-gh-nkumar** [comment]
> 1. Thanks for the PR with informative description.  QQ:  the description said " It's possible that lambda passed in transform does not adhere to expected output schema. It should be rare", what is the behavior when the lambda passed does not adhere to expected output schema?
> 2. why all the precommit tests failed?

1. It will fail with due to UDTF schema mismatch. But I plan support this edge case as well in follow up PR.
2. It's just branch is out-of-date. Need to rebase on latest main.

**Comment by sfc-gh-nkumar** [comment]
> This looks pretty good to me. What happens with `test_dataframe_groupby_transform_conflicting_labels_negative` - does it just not throw an exception or what?

New implementation does not raise NotImplementedError as expected by test but Schema mismatch error. Planning to fix this in follow up PR.

**Comment by sfc-gh-nkumar** [comment]
> > 3. Note: It's possible that lambda passed in transform does not adhere to expected output schema. It should be rare but will improve this as well when we add type inference in groupby.transform.
> 
> I think it would be good if we have a test to exercise this case.

Done

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM, thanks Naresh!

**Review by sfc-gh-yuwang — APPROVED** [review]
LGTM! just make sure all test passed before merge, thanks!
